operationsgateway_api/src/auth/oidc_handler.py

```diff
@@ -0,0 +1,136 @@
+            jwks_config = r.json()
+        except Exception as exc:
+            log.exception("Failed to fetch JWKS from %s:", jwks_uri)
+            raise AuthServerError() from exc
+
+        self._keys = {}
+        for key in jwks_config.get("keys", []):
+            kid = key.get("kid")
+            if key.get("use") != "sig":
+                log.debug("Skipping non-signing key %s (use=%s)", kid, key.get("use"))
+                continue
+            try:
+                self._keys[kid] = jwt.PyJWK(key)
+            except (jwt.exceptions.PyJWKError, jwt.exceptions.InvalidKeyError):
+                log.warning("Could not load key")
+
+    def get_issuer(self) -> str:
+        return self._issuer
+
+    def get_key(self, kid: str) -> jwt.PyJWK:
+        if kid not in self._keys:
+            log.exception("Unknown key ID %s:", kid)
+            raise AuthServerError()
+        return self._keys[kid]
+
+    def get_audience(self) -> str:
+        return self._audience
+
+    def get_mechanism(self) -> str:
+        return self._mechanism
+
+    def get_matching_claim(self) -> str:
+        return self._matching_claim
+
+
+class OidcHandler:
+    def __init__(self) -> None:
+        self._providers = {}
+
+        for name, provider_config in Config.config.auth.oidc_providers.items():
+            log.info("Loading OIDC provider: %s", name)
+            try:
+                provider = OidcProvider(provider_config)
+                self._providers[provider.get_issuer()] = provider
+            except Exception as exc:
+                log.error("Initialisation error for OIDC provider %s", name)
+                raise AuthServerError() from exc
+
+    def handle(self, encoded_token: str) -> str:
+        try:
+            # Decode header/payload to extract issuer and kid
+            unverified_header = jwt.get_unverified_header(encoded_token)
+            unverified_payload = jwt.decode(
+                encoded_token,
+                options={"verify_signature": False},
+            )
+
+            kid = unverified_header["kid"]
+            iss = unverified_payload["iss"]
+            if iss not in self._providers:
+                log.error("Unknown issuer: %s", iss)
+                raise AuthServerError()
+
+            provider = self._providers[iss]
+            key = provider.get_key(kid)
+
+            payload = jwt.decode(
+                encoded_token,
+                key=key,
+                algorithms=[key.algorithm_name],
+                audience=provider.get_audience(),
+                options={
+                    "require": ["exp", "aud"],
+                    "verify_exp": True,
+                    "verify_aud": True,
+                },
+            )
+
+            matching_claim = payload.get(provider.get_matching_claim())
+            if not matching_claim:
+                log.error("Username claim missing in ID token")
+                raise UnauthorisedError("Username claim missing in ID token")
```

Comment:
```suggestion
            matching_claim_key = provider.get_matching_claim()
            matching_claim = payload.get(matching_claim_key)
            if not matching_claim:
                log.error("%s claim missing in ID token", matching_claim_key)
                raise UnauthorisedError(f"{matching_claim_key} claim missing in ID token")
```

As I understand this, in general we could be looking for any key here but in practice we are looking for `"email"`. I'm wondering if perhaps the log/error should refer to this rather than `"Username"` (and maybe the reference to username is a hangover from when we were planning to use fedId rather than email to do this)? I may have misunderstood so happy for this to stay as is (and my suggested change might not run as is anyway).

Reply:
Done. Yeah this was a hangover from the first pass implementation. In some providers, the email is returned in the username key, so I was trying to remain username/email agnostic.
However, we'll just be dealing with email, so we should really just use that. We could probably get rid of the `matching_claim` in the config as well, but I'm happy to leave that in for now until its running in prod.

Reply:
Also, i've put the STFC Tenacy details in a secure note in Keeper